tensorflow/python/kernel_tests/xent_op_test_base.py

```diff
@@ -0,0 +1,362 @@
+# Copyright 2015-2021 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Tests for SoftmaxCrossEntropyWithLogits op."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import os
+
+import numpy as np
+
+from tensorflow.python.framework import config
+from tensorflow.python.framework import constant_op
+from tensorflow.python.framework import dtypes
+from tensorflow.python.framework import test_util
+from tensorflow.python.ops import array_ops
+from tensorflow.python.ops import gen_nn_ops
+from tensorflow.python.ops import gradient_checker
+from tensorflow.python.ops import gradients_impl
+from tensorflow.python.ops import math_ops
+from tensorflow.python.ops import nn_ops
+# The following import is required to register the gradient function.
+from tensorflow.python.ops.nn_grad import _SoftmaxCrossEntropyWithLogitsGrad  # pylint: disable=unused-import
+from tensorflow.python.platform import test
+
+
+class XentOpTestBase(test.TestCase):
+
+  def _opDeterminismEnabled(self):
+    deterministic_ops = os.getenv('TF_DETERMINISTIC_OPS', '0')
+    return deterministic_ops == '1' or deterministic_ops == 'true'
+
+  def _coreOp(self, features, labels):
```

Comment:
I think instead of having a `_coreOp` function, we should just test the `nn_ops` version instead of the `gen_nn_ops` version for most tests. This avoids the awkwardness of having to create a Python implementation that has the same interface as the C++ op in `XentDeterministicTest`.

For test methods that are specific to the `gen_nn_ops` version, like `testRankTooLarge`, you can define a `XentOpTest` class subclassing `XentOpTestBase` and put them there. That way, you don't have to override methods in `XentDeterministicTest` to do nothing.

Once we have a flag to enable determinism instead of the environmental variable, we can have just a single test file and a `testWithDeterminismOnAndOff` test decorator, which will simply things. But we're not quite there yet.

Reply:
> I think instead of having a `_coreOp` function, we should just test the `nn_ops` version instead of the `gen_nn_ops` version for most tests. This avoids the awkwardness of having to create a Python implementation that has the same interface as the C++ op in `XentDeterministicTest`.

Sounds good. I was trying to minimize changes and be done quicker. What you propose is a good technical debt minimization step.

> For test methods that are specific to the `gen_nn_ops` version, like `testRankTooLarge`, you can define a `XentOpTest` class subclassing `XentOpTestBase` and put them there. That way, you don't have to override methods in `XentDeterministicTest` to do nothing.

Yep. Sounds good.

> Once we have a flag to enable determinism instead of the environmental variable, we can have just a single test file and a `testWithDeterminismOnAndOff` test decorator, which will simply things. But we're not quite there yet.

Absolutely. The (temporary) environment variable implementation makes organization of test files and classes unusually cumbersome. Looking forward to the ensuing tidiness.